Fix #43815: Improve error message for resource functions

oshi-git opened the pull request:
## Description
Fixes #43815

This PR improves the error message when resource functions have incomplete return handling. The error message now correctly displays "resource function" instead of the confusing "resource_func".

## Problem
When resource functions had incomplete return handling in `on fail` blocks, the compiler would show:
```
error: this resource_func must return a result
```
This wasn't very clear because:
- `resource_func` is not valid Ballerina syntax
- The correct syntax is `resource function` (two words)
- Users didn't understand what "resource_func" meant

## Solution
- Modified `ReachabilityAnalyzer.java` to format function kind names properly
- Added `getFunctionKindName()` helper method that converts:
  - `"resource_func"` → `"resource function"`
  - Other function kinds remain unchanged
- Error message now clearly indicates the correct syntax

## Changes Made
**Modified:** `compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java`
- Line 584: Changed to use `getFunctionKindName(funcNode)` instead of `funcNode.getKind().toString().toLowerCase()`
- Lines 1034-1040: Added new helper method `getFunctionKindName()`

## Before
```error: this resource_func must return a result
```

## After
```error: this resource function must return a result
```

## Testing
- Manually tested with resource functions that have incomplete return handling
- Error message now displays correctly
- No impact on other function types

## Checklist
- [x] Code follows Ballerina style guidelines
- [x] Fixes the reported issue
- [x] No breaking changes
- [x] Linked to issue #43815

dulajdilshan commented:
@oshi-git The test you have added is not properly executing. Please check how other negative tests have been added.

oshi-git commented:
@dulajdilshan Thank you for the guidance! I've now added the Java test case:

**Changes:**
- Added `testResourceFunctionMustReturn()` method to `ReachabilityAnalysisTest.java`
- The test properly compiles and validates the `.bal` test file
- Follows the same pattern as other negative tests in the file (like `testUnreachability()`)
- Validates that the error message correctly displays "this resource function must return a result" on line 24

The test should now execute properly. Please let me know if any adjustments are needed!

Thank you for your patience in helping me understand the test structure!

MaryamZi commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java:
```diff
@@ -1031,4 +1032,11 @@ public static class AnalyzerData {
+            return "resource function";
+    	}
+        return kind.toString().toLowerCase();
+    }
```
```suggestion

private String getFunctionKindName(BLangFunction funcNode) {
    NodeKind kind = funcNode.getKind();
    if (kind == NodeKind.RESOURCE_FUNC) {
        return "resource function";
    }
    return kind.toString().toLowerCase();
}

```

MaryamZi commented on tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/resource_function_must_return_negative.bal:
```diff
@@ -0,0 +1,30 @@
+
+function getNumber() returns int|error {
+    return error("test error");
+}
```
```suggestion
}

```

MaryamZi commented on tests/jballerina-unit-test/src/test/resources/test-src/reachability-analysis/resource_function_must_return_negative.bal:
```diff
@@ -0,0 +1,30 @@
+    resource function get testPath() returns int|error {
+        do {
+            int x = check getNumber();
+	    return x;
```
```suggestion
            return x;
```

oshi-git commented:
@MaryamZi Thank you for catching those formatting issues! I've applied all three formatting suggestions:

**Changes applied:**
1. ✅ Fixed indentation in Java file method
2. ✅ Added newline at end of test file
3. ✅ Fixed tab character in return statement

Thank you for your patience and guidance! 🙏

keizer619 commented:
@oshi-git There seems to be compilation falures. Hence i will be closing the PR, feel free to fix and reopen
